src/backend/distributed/planner/fast_path_router_planner.c

```diff
@@ -0,0 +1,391 @@
+ * The returned PlannedStmt is not proper (e.g., set_plan_references() is
+ * not called on the plan or the quals are not set), so should not be
+ * passed to the executor directly. This is only useful to have a
+ * placeholder PlannedStmt where target list is properly set. Note that
+ * this is what router executor relies on.
+ *
+ * This function makes the assumption (and the assertion) that
+ * the input query is in the form defined by FastPathRouterQuery().
+ */
+PlannedStmt *
+GeneratePlaceHolderPlannedStmt(Query *parse)
+{
+	PlannedStmt *result = makeNode(PlannedStmt);
+	SeqScan *seqScanNode = makeNode(SeqScan);
+	Plan *plan = &seqScanNode->plan;
+	Oid relationId = InvalidOid;
+
+	AssertArg(FastPathRouterQuery(parse));
+
+	/* there is only a single relation rte */
+	seqScanNode->scanrelid = 1;
+
+	plan->targetlist = copyObject(parse->targetList);
+	plan->qual = NULL;
+	plan->lefttree = NULL;
+	plan->righttree = NULL;
+	plan->plan_node_id = 1;
+
+	/*  rtable is used for access permission checks */
+	result->commandType = parse->commandType;
+	result->queryId = parse->queryId;
+	result->stmt_len = parse->stmt_len;
+
+	result->rtable = copyObject(parse->rtable);
+	result->planTree = (Plan *) plan;
+
+	relationId = ExtractFirstDistributedTableId(parse);
+	result->relationOids = list_make1_oid(relationId);
+
+	return result;
+}
+
+
+/*
+ * FastPathRouterQuery gets a query and returns true if the query is eligable for
+ * being a fast path router query.
+ * The requirements for the fast path query can be listed below:
+ *
+ *   - SELECT query without CTES, sublinks-subqueries, set operations
+ *   - The query should touch only a single hash distributed or reference table
+ *   - The distribution with equality operator should be in the WHERE clause
+ *      and it should be ANDed with any other filters. Also, the distribution
+ *      key should only exists once in the WHERE clause. So basically,
+ *          SELECT ... FROM dist_table WHERE dist_key = X
+ *   - No returning for UPDATE/DELETE queries
+ */
+bool
+FastPathRouterQuery(Query *query)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	List *rangeTableList = NIL;
+	FromExpr *joinTree = query->jointree;
+	Node *quals = NULL;
+	Oid distributedTableId = InvalidOid;
+	Var *distributionKey = NULL;
+	DistTableCacheEntry *cacheEntry = NULL;
+	List *varClauseList = NIL;
+	ListCell *varClauseCell = NULL;
+	int partitionColumnReferenceCount = 0;
+
+	if (!EnableFastPathRouterPlanner)
+	{
+		return false;
+	}
+
+	if (!(query->commandType == CMD_SELECT || query->commandType == CMD_UPDATE ||
+		  query->commandType == CMD_DELETE))
+	{
+		return false;
+	}
+
+	/*
+	 * We want to deal with only very simple select queries. Some of the
+	 * checks might be too restrictive, still we prefer this way.
+	 */
+	if (query->cteList != NIL || query->returningList != NIL ||
+		query->hasSubLinks || query->setOperations != NULL ||
+		query->hasForUpdate || query->hasTargetSRFs ||
```

Comment:
worth confirming whether supporting FOR UPDATE makes a difference in benchmarks

Reply:
For pgbench tests it didn't have any affect (as expected). I'll double check with TPC-C tests, which I'll run next.

Reply:
So, it is clear that allowing `hasForUpdate` improves the performance and we should do it.

One edge case is that `MultiRouterPlannableQuery ` usees restrictInfos to disable replicationFactor > 1 and hasForUpdate (except for reference table). 

If I simply remove `query->hasForUpdate` from there, we'd silently allow it.

So, I'd either (a) disable query->hasForUpdate with replication factor > 1 here, (b) refactor `MultiRouterPlannableQuery` which might require query traversal (e.g., call `ExtractRangeTableRelationWalker`), (c) enable hasForUpdate with replication > 1 which I need to look a bit and probably would take more time.

Any preference @marcocitus ?

Reply:
> (a) disable query->hasForUpdate with replication factor > 1 here,

For now, I pushed this.

Reply:
Looking at this again. I'd have a slight preference for refactoring `MultiRouterPlannableQuery`, because we're quietly skipping a bunch of checks which will not be obvious to readers of that function and may lead to bugs down the road.